### PR TITLE
IKEv2: fix length calculation for IKEv2_Notify payloads with SPI

### PR DESCRIPTION
**Pull request by mspncp**
As a testcase, add a decrypted CREATE_CHILD_SA request for an ESP child sa, which contains a REKEY_SA Notify payload.

### CODE REVIEW

**Comment by mspncp** [comment]
# Example

The following example shows how the decrypted CREATE_CHILD_SA request is dissected incorrectly on master and how it is dissected with the change in this pull request:

```python
#!/usr/bin/env python3

from scapy.all import *
from scapy.contrib.ikev2 import *

import binascii

decrypted_frame = binascii.unhexlify(''.join("""
00 50 56 99 bf d5 00 50  56 99 69 93 08 00 45 00
01 38 60 32 40 00 40 11  c1 0f 0a 05 02 36 0a 05
02 34 b8 99 11 94 01 24  19 a9 00 00 00 00 46 b3
f6 88 4d 37 5f 9a f5 38  82 35 ea 87 5e 8a 29 20
24 00 00 00 00 00 00 00  01 18            

21 00 00 0c 03 04 40 09  5f c7 ff 5a 28 00 00 2c
00 00 00 28 01 03 04 03  6b 21 88 20 03 00 00 0c
01 00 00 14 80 0e 00 80  03 00 00 08 04 00 00 1c
00 00 00 08 05 00 00 00  22 00 00 2c ea 7e 88 57
4a 36 64 cd 67 e3 3c 42  46 66 59 4d df 70 25 03
b2 00 a3 3f 87 82 f2 3c  94 c0 60 0e ae 7e d9 50
d7 67 e9 6e 2c 00 00 48  00 1c 00 00 8e 15 b1 f4
9a cc 04 ff 12 e3 2f bc  3a f0 57 14 81 f3 b9 6c
21 1a f7 36 97 6d c2 23  80 74 ef 75 59 d1 99 65
5a a5 80 00 87 4a bf 1f  13 f7 e1 6f de 34 80 94
28 1c 93 cb 5a ee 30 24  d9 3e b9 55 2d 00 00 18
01 00 00 00 07 00 00 10  00 00 ff ff c0 a8 e1 0b
c0 a8 e1 0b 00 00 00 18  01 00 00 00 07 00 00 10
00 00 ff ff c0 a8 e1 00  c0 a8 e1 ff            
""".split()))

Ether(decrypted_frame)['IKEv2'].show()
```

### Broken (master)

```
###[ IKEv2 ]### 
  init_SPI  = 46b3f6884d375f9a
  resp_SPI  = f5388235ea875e8a
  next_payload= Notify
  version   = 0x20
  exch_type = CREATE_CHILD_SA
  flags     = 
  id        = 0
  length    = 280
###[ IKEv2 Notify ]### 
     next_payload= SA
     flags     = 
     length    = 12
     proto     = ESP
     SPIsize   = 4
     type      = REKEY_SA
     SPI       = 5fc7ff5a
     notify    = 2800002c
###[ IKEv2 SA ]### 
        next_payload= None
        flags     = 
        length    = 40
        \prop      \
         |###[ IKEv2 Proposal ]### 
         |  next_payload= 1
         |  flags     = critical+?
         |  length    = 1027
         |  proposal  = 107
         |  proto     = 33
         |  SPIsize   = 136
         |  trans_nb  = 32
         |  SPI       = 0300000c01000014800e0080030000080400001c0000000805000000
         |  \trans     \
         |   |###[ Raw ]### 
         |   |  load      = ''
###[ Raw ]### 
           load      = '"\x00\x00,\\xea~\\x88WJ6d\\xcdg\\xe3<BFfYM\\xdfp%\x03\\xb2\x00\\xa3?\\x87\\x82\\xf2<\\x94\\xc0`\x0e\\xae~\\xd9P\\xd7g\\xe9n,\x00\x00H\x00\x1c\x00\x00\\x8e\x15\\xb1\\xf4\\x9a\\xcc\x04\\xff\x12\\xe3/\\xbc:\\xf0W\x14\\x81\\xf3\\xb9l!\x1a\\xf76\\x97m\\xc2#\\x80t\\xefuYљeZ\\xa5\\x80\x00\\x87J\\xbf\x1f\x13\\xf7\\xe1o\\xde4\\x80\\x94(\x1c\\x93\\xcbZ\\xee0$\\xd9>\\xb9U-\x00\x00\x18\x01\x00\x00\x00\x07\x00\x00\x10\x00\x00\\xff\\xff\\xc0\\xa8\\xe1\x0b\\xc0\\xa8\\xe1\x0b\x00\x00\x00\x18\x01\x00\x00\x00\x07\x00\x00\x10\x00\x00\\xff\\xff\\xc0\\xa8\\xe1\x00\\xc0\\xa8\\xe1\\xff'
```



### Fixed (this pull request)

```
###[ IKEv2 ]### 
  init_SPI  = 46b3f6884d375f9a
  resp_SPI  = f5388235ea875e8a
  next_payload= Notify
  version   = 0x20
  exch_type = CREATE_CHILD_SA
  flags     = 
  id        = 0
  length    = 280
###[ IKEv2 Notify ]### 
     next_payload= SA
     flags     = 
     length    = 12
     proto     = ESP
     SPIsize   = 4
     type      = REKEY_SA
     SPI       = 5fc7ff5a
     notify    = ###[ IKEv2 SA ]### 
        next_payload= Nonce
        flags     = 
        length    = 44
        \prop      \
         |###[ IKEv2 Proposal ]### 
         |  next_payload= None
         |  flags     = 
         |  length    = 40
         |  proposal  = 1
         |  proto     = ESP
         |  SPIsize   = 4
         |  trans_nb  = 3
         |  SPI       = 6b218820
         |  \trans     \
         |   |###[ IKEv2 Transform ]### 
         |   |  next_payload= Transform
         |   |  flags     = 
         |   |  length    = 12
         |   |  transform_type= Encryption
         |   |  res2      = 0
         |   |  transform_id= AES-GCM-16ICV
         |   |  key_length= 128
         |   |###[ IKEv2 Transform ]### 
         |   |     next_payload= Transform
         |   |     flags     = 
         |   |     length    = 8
         |   |     transform_type= GroupDesc
         |   |     res2      = 0
         |   |     transform_id= brainpoolP256r1gr
         |   |###[ IKEv2 Transform ]### 
         |   |        next_payload= None
         |   |        flags     = 
         |   |        length    = 8
         |   |        transform_type= Extended Sequence Number
         |   |        res2      = 0
         |   |        transform_id= No ESN
###[ IKEv2 Nonce ]### 
           next_payload= KE
           flags     = 
           length    = 44
           nonce     = ea7e88574a3664cd67e33c424666594ddf702503b200a33f8782f23c94c0600eae7ed950d767e96e
###[ IKEv2 Key Exchange ]### 
              next_payload= TSi
              flags     = 
              length    = 72
              group     = brainpoolP256r1gr
              res2      = 0
              ke        = 8e15b1f49acc04ff12e32fbc3af0571481f3b96c211af736976dc2238074ef7559d199655aa58000874abf1f13f7e16fde348094281c93cb5aee3024d93eb955
###[ IKEv2 Traffic Selector - Initiator ]### 
                 next_payload= TSr
                 flags     = 
                 length    = 24
                 number_of_TSs= 1
                 res2      = 0x0
                 \traffic_selector\
                  |###[ IKEv2 IPv4 Traffic Selector ]### 
                  |  TS_type   = TS_IPV4_ADDR_RANGE
                  |  IP_protocol_ID= All protocols
                  |  length    = 16
                  |  start_port= 0
                  |  end_port  = 65535
                  |  starting_address_v4= 192.168.225.11
                  |  ending_address_v4= 192.168.225.11
###[ IKEv2 Traffic Selector - Responder ]### 
                    next_payload= None
                    flags     = 
                    length    = 24
                    number_of_TSs= 1
                    res2      = 0x0
                    \traffic_selector\
                     |###[ IKEv2 IPv4 Traffic Selector ]### 
                     |  TS_type   = TS_IPV4_ADDR_RANGE
                     |  IP_protocol_ID= All protocols
                     |  length    = 16
                     |  start_port= 0
                     |  end_port  = 65535
                     |  starting_address_v4= 192.168.225.0
                     |  ending_address_v4= 192.168.225.255

```


### TShark

Here is the TShark dump of the original capture  (with decryption secrets available)

**tshark -r CREATE_CHILD_SA.pcapng -V -x**

```
Frame 1: 326 bytes on wire (2608 bits), 326 bytes captured (2608 bits) on interface eth_ext, id 2
    Section number: 1
    Interface id: 2 (eth_ext)
        Interface name: eth_ext
    Encapsulation type: Ethernet (1)
    Arrival Time: May  2, 2023 17:41:42.378369439 CEST
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1683042102.378369439 seconds
    [Time delta from previous captured frame: 0.000000000 seconds]
    [Time delta from previous displayed frame: 0.000000000 seconds]
    [Time since reference or first frame: 0.000000000 seconds]
    Frame Number: 1
    Frame Length: 326 bytes (2608 bits)
    Capture Length: 326 bytes (2608 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: eth:ethertype:ip:udp:udpencap:isakmp]
Ethernet II, Src: VMware_99:69:93 (00:50:56:99:69:93), Dst: VMware_99:bf:d5 (00:50:56:99:bf:d5)
    Destination: VMware_99:bf:d5 (00:50:56:99:bf:d5)
        Address: VMware_99:bf:d5 (00:50:56:99:bf:d5)
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Source: VMware_99:69:93 (00:50:56:99:69:93)
        Address: VMware_99:69:93 (00:50:56:99:69:93)
        .... ..0. .... .... .... .... = LG bit: Globally unique address (factory default)
        .... ...0 .... .... .... .... = IG bit: Individual address (unicast)
    Type: IPv4 (0x0800)
Internet Protocol Version 4, Src: 10.5.2.54, Dst: 10.5.2.52
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
        0000 00.. = Differentiated Services Codepoint: Default (0)
        .... ..00 = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    Total Length: 312
    Identification: 0x6032 (24626)
    010. .... = Flags: 0x2, Don't fragment
        0... .... = Reserved bit: Not set
        .1.. .... = Don't fragment: Set
        ..0. .... = More fragments: Not set
    ...0 0000 0000 0000 = Fragment Offset: 0
    Time to Live: 64
    Protocol: UDP (17)
    Header Checksum: 0xc10f [validation disabled]
    [Header checksum status: Unverified]
    Source Address: 10.5.2.54
    Destination Address: 10.5.2.52
User Datagram Protocol, Src Port: 47257, Dst Port: 4500
    Source Port: 47257
    Destination Port: 4500
    Length: 292
    Checksum: 0x19a9 [unverified]
    [Checksum Status: Unverified]
    [Stream index: 0]
    [Timestamps]
        [Time since first frame: 0.000000000 seconds]
        [Time since previous frame: 0.000000000 seconds]
    UDP payload (284 bytes)
UDP Encapsulation of IPsec Packets
    Non-ESP Marker
Internet Security Association and Key Management Protocol
    Initiator SPI: 46b3f6884d375f9a
    Responder SPI: f5388235ea875e8a
    Next payload: Encrypted and Authenticated (46)
    Version: 2.0
        0010 .... = MjVer: 0x2
        .... 0000 = MnVer: 0x0
    Exchange type: CREATE_CHILD_SA (36)
    Flags: 0x00 (Responder, No higher version, Request)
        .... 0... = Initiator: Responder
        ...0 .... = Version: No higher version
        ..0. .... = Response: Request
    Message ID: 0x00000000
    Length: 280
    Payload: Encrypted and Authenticated (46)
        Next payload: Notify (41)
        0... .... = Critical Bit: Not critical
        .000 0000 = Reserved: 0x00
        Payload length: 252
        Initialization Vector: 4353b6d55744ae42 (8 bytes)
        Encrypted Data (224 bytes) <AES-GCM-256 with 16 octet ICV [RFC5282]>
        Decrypted Data (224 bytes)
            Contained Data (220 bytes)
                Payload: Notify (41) - REKEY_SA
                    Next payload: Security Association (33)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 12
                    Protocol ID: ESP (3)
                    SPI Size: 4
                    Notify Message Type: REKEY_SA (16393)
                    SPI: 5fc7ff5a
                    Notification DATA: <MISSING>
                Payload: Security Association (33)
                    Next payload: Nonce (40)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 44
                    Payload: Proposal (2) # 1
                        Next payload: NONE / No Next Payload  (0)
                        Reserved: 00
                        Payload length: 40
                        Proposal number: 1
                        Protocol ID: ESP (3)
                        SPI Size: 4
                        Proposal transforms: 3
                        SPI: 6b218820
                        Payload: Transform (3)
                            Next payload: Transform (3)
                            Reserved: 00
                            Payload length: 12
                            Transform Type: Encryption Algorithm (ENCR) (1)
                            Reserved: 00
                            Transform ID (ENCR): AES-GCM with a 16 octet ICV (20)
                            Transform Attribute (t=14,l=2): Key Length: 128
                                1... .... .... .... = Format: Type/Value (TV)
                                Type: Key Length (14)
                                Value: 0080
                                Key Length: 128
                        Payload: Transform (3)
                            Next payload: Transform (3)
                            Reserved: 00
                            Payload length: 8
                            Transform Type: Diffie-Hellman Group (D-H) (4)
                            Reserved: 00
                            Transform ID (D-H): 256-bit Brainpool ECP group (28)
                        Payload: Transform (3)
                            Next payload: NONE / No Next Payload  (0)
                            Reserved: 00
                            Payload length: 8
                            Transform Type: Extended Sequence Numbers (ESN) (5)
                            Reserved: 00
                            Transform ID (ESN): No Extended Sequence Numbers (0)
                Payload: Nonce (40)
                    Next payload: Key Exchange (34)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 44
                    Nonce DATA: ea7e88574a3664cd67e33c424666594ddf702503b200a33f8782f23c94c0600eae7ed950…
                Payload: Key Exchange (34)
                    Next payload: Traffic Selector - Initiator (44)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 72
                    DH Group #: 256-bit Brainpool ECP group (28)
                    Reserved: 0000
                    Key Exchange Data: 8e15b1f49acc04ff12e32fbc3af0571481f3b96c211af736976dc2238074ef7559d19965…
                Payload: Traffic Selector - Initiator (44) # 1
                    Next payload: Traffic Selector - Responder (45)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 24
                    Number of Traffic Selectors: 1
                    Reserved: 000000
                    Traffic Selector: TS_IPV4_ADDR_RANGE
                        Traffic Selector Type: TS_IPV4_ADDR_RANGE (7)
                        Protocol ID: Unused
                        Selector Length: 16
                        Start Port: 0
                        End Port: 65535
                        Starting Addr: 192.168.225.11
                        Ending Addr: 192.168.225.11
                Payload: Traffic Selector - Responder (45) # 1
                    Next payload: NONE / No Next Payload  (0)
                    0... .... = Critical Bit: Not critical
                    .000 0000 = Reserved: 0x00
                    Payload length: 24
                    Number of Traffic Selectors: 1
                    Reserved: 000000
                    Traffic Selector: TS_IPV4_ADDR_RANGE
                        Traffic Selector Type: TS_IPV4_ADDR_RANGE (7)
                        Protocol ID: Unused
                        Selector Length: 16
                        Start Port: 0
                        End Port: 65535
                        Starting Addr: 192.168.225.0
                        Ending Addr: 192.168.225.255
            Padding (3 bytes)
            Pad Length: 3
        Integrity Checksum Data: 03d4e2681957b4e5bd0a95ed2beb5912 (16 bytes)[correct]

Frame (326 bytes):
0000  00 50 56 99 bf d5 00 50 56 99 69 93 08 00 45 00   .PV....PV.i...E.
0010  01 38 60 32 40 00 40 11 c1 0f 0a 05 02 36 0a 05   .8`2@.@......6..
0020  02 34 b8 99 11 94 01 24 19 a9 00 00 00 00 46 b3   .4.....$......F.
0030  f6 88 4d 37 5f 9a f5 38 82 35 ea 87 5e 8a 2e 20   ..M7_..8.5..^.. 
0040  24 00 00 00 00 00 00 00 01 18 29 00 00 fc 43 53   $.........)...CS
0050  b6 d5 57 44 ae 42 61 0b 18 83 1a 69 4b 67 60 5c   ..WD.Ba....iKg`\
0060  46 4f 78 8e 77 83 0d 3f 82 8d c2 0f 62 38 b4 0e   FOx.w..?....b8..
0070  45 1f fc ba 6c 88 e3 9a f1 bb 8b df 47 8f b1 35   E...l.......G..5
0080  22 42 15 c5 1d 76 64 a5 8f 6f ec 18 ad cd ad 2d   "B...vd..o.....-
0090  a2 c5 8e 47 1d 4c ce 76 dd de 43 5f 39 80 d6 b4   ...G.L.v..C_9...
00a0  9c 02 71 b9 37 a7 a3 a2 cf 48 54 6a 47 6e 6f ea   ..q.7....HTjGno.
00b0  60 c8 be e5 42 8b f8 f5 70 fa 7e fc e1 f9 f8 6d   `...B...p.~....m
00c0  33 6e d6 ce e3 cb 7c dd 62 dc 6a 02 86 07 8b 60   3n....|.b.j....`
00d0  05 57 bc 73 18 3d 1f 8f da 38 07 f8 6b 8e 1a 16   .W.s.=...8..k...
00e0  91 b1 e6 54 b6 99 91 b2 e9 4d fa 6e c4 0a 17 53   ...T.....M.n...S
00f0  38 a0 97 f1 81 7f 87 53 ab 66 c7 57 5e 1f 36 28   8......S.f.W^.6(
0100  af e1 94 59 8b d3 aa 51 64 9f 28 a8 da 35 6a a2   ...Y...Qd.(..5j.
0110  b8 43 e9 f7 c4 ff 2e 7a ef e3 b5 b5 44 29 ee 02   .C.....z....D)..
0120  76 93 b3 33 ec 44 f8 43 1b a3 34 d4 2e 52 67 c7   v..3.D.C..4..Rg.
0130  61 02 94 03 73 33 03 d4 e2 68 19 57 b4 e5 bd 0a   a...s3...h.W....
0140  95 ed 2b eb 59 12                                 ..+.Y.
Decrypted Data (224 bytes):
0000  21 00 00 0c 03 04 40 09 5f c7 ff 5a 28 00 00 2c   !.....@._..Z(..,
0010  00 00 00 28 01 03 04 03 6b 21 88 20 03 00 00 0c   ...(....k!. ....
0020  01 00 00 14 80 0e 00 80 03 00 00 08 04 00 00 1c   ................
0030  00 00 00 08 05 00 00 00 22 00 00 2c ea 7e 88 57   ........"..,.~.W
0040  4a 36 64 cd 67 e3 3c 42 46 66 59 4d df 70 25 03   J6d.g.<BFfYM.p%.
0050  b2 00 a3 3f 87 82 f2 3c 94 c0 60 0e ae 7e d9 50   ...?...<..`..~.P
0060  d7 67 e9 6e 2c 00 00 48 00 1c 00 00 8e 15 b1 f4   .g.n,..H........
0070  9a cc 04 ff 12 e3 2f bc 3a f0 57 14 81 f3 b9 6c   ....../.:.W....l
0080  21 1a f7 36 97 6d c2 23 80 74 ef 75 59 d1 99 65   !..6.m.#.t.uY..e
0090  5a a5 80 00 87 4a bf 1f 13 f7 e1 6f de 34 80 94   Z....J.....o.4..
00a0  28 1c 93 cb 5a ee 30 24 d9 3e b9 55 2d 00 00 18   (...Z.0$.>.U-...
00b0  01 00 00 00 07 00 00 10 00 00 ff ff c0 a8 e1 0b   ................
00c0  c0 a8 e1 0b 00 00 00 18 01 00 00 00 07 00 00 10   ................
00d0  00 00 ff ff c0 a8 e1 00 c0 a8 e1 ff 01 02 03 03   ................
```

### SCapy

```python
from scapy.all import *
from scapy.contrib.ikev2 import *
with open('example/CREATE_CHILD_SA.decrypted.raw', 'rb') as f:
     packet=f.read()

print("Decrypted Data")
hexdump(packet[28:])
```

```
Decrypted Data
0000  21 00 00 0C 03 04 40 09 5F C7 FF 5A 28 00 00 2C  !.....@._..Z(..,
0010  00 00 00 28 01 03 04 03 6B 21 88 20 03 00 00 0C  ...(....k!. ....
0020  01 00 00 14 80 0E 00 80 03 00 00 08 04 00 00 1C  ................
0030  00 00 00 08 05 00 00 00 22 00 00 2C EA 7E 88 57  ........"..,.~.W
0040  4A 36 64 CD 67 E3 3C 42 46 66 59 4D DF 70 25 03  J6d.g.<BFfYM.p%.
0050  B2 00 A3 3F 87 82 F2 3C 94 C0 60 0E AE 7E D9 50  ...?...<..`..~.P
0060  D7 67 E9 6E 2C 00 00 48 00 1C 00 00 8E 15 B1 F4  .g.n,..H........
0070  9A CC 04 FF 12 E3 2F BC 3A F0 57 14 81 F3 B9 6C  ....../.:.W....l
0080  21 1A F7 36 97 6D C2 23 80 74 EF 75 59 D1 99 65  !..6.m.#.t.uY..e
0090  5A A5 80 00 87 4A BF 1F 13 F7 E1 6F DE 34 80 94  Z....J.....o.4..
00a0  28 1C 93 CB 5A EE 30 24 D9 3E B9 55 2D 00 00 18  (...Z.0$.>.U-...
00b0  01 00 00 00 07 00 00 10 00 00 FF FF C0 A8 E1 0B  ................
00c0  C0 A8 E1 0B 00 00 00 18 01 00 00 00 07 00 00 10  ................
00d0  00 00 FF FF C0 A8 E1 00 C0 A8 E1 FF              ............

###[ IKEv2 ]### 
  init_SPI  = 46b3f6884d375f9a
  resp_SPI  = f5388235ea875e8a
  next_payload= Notify
  version   = 0x20
  exch_type = CREATE_CHILD_SA
  flags     = 
  id        = 0
  length    = 280
###[ IKEv2 Notify ]### 
     next_payload= SA
     flags     = 
     length    = 12
     proto     = ESP
     SPIsize   = 4
     type      = REKEY_SA
     SPI       = 5fc7ff5a
     notify    = 2800002c
###[ IKEv2 SA ]### 
        next_payload= None
        flags     = 
        length    = 40
        \prop      \
         |###[ IKEv2 Proposal ]### 
         |  next_payload= 1
         |  flags     = critical+?
         |  length    = 1027
         |  proposal  = 107
         |  proto     = 33
         |  SPIsize   = 136
         |  trans_nb  = 32
         |  SPI       = 0300000c01000014800e0080030000080400001c0000000805000000
         |  \trans     \
         |   |###[ Raw ]### 
         |   |  load      = ''
###[ Raw ]### 
           load      = '"\x00\x00,\\xea~\\x88WJ6d\\xcdg\\xe3<BFfYM\\xdfp%\x03\\xb2\x00\\xa3?\\x87\\x82\\xf2<\\x94\\xc0`\x0e\\xae~\\xd9P\\xd7g\\xe9n,\x00\x00H\x00\x1c\x00\x00\\x8e\x15\\xb1\\xf4\\x9a\\xcc\x04\\xff\x12\\xe3/\\xbc:\\xf0W\x14\\x81\\xf3\\xb9l!\x1a\\xf76\\x97m\\xc2#\\x80t\\xefuYљeZ\\xa5\\x80\x00\\x87J\\xbf\x1f\x13\\xf7\\xe1o\\xde4\\x80\\x94(\x1c\\x93\\xcbZ\\xee0$\\xd9>\\xb9U-\x00\x00\x18\x01\x00\x00\x00\x07\x00\x00\x10\x00\x00\\xff\\xff\\xc0\\xa8\\xe1\x0b\\xc0\\xa8\\xe1\x0b\x00\x00\x00\x18\x01\x00\x00\x00\x07\x00\x00\x10\x00\x00\\xff\\xff\\xc0\\xa8\\xe1\x00\\xc0\\xa8\\xe1\\xff'
```

File: scapy/contrib/ikev2.py [hunk 1]
```diff
@@ -748,7 +748,7 @@ class IKEv2_Notify(IKEv2_Payload):
         ShortEnumField("type", 0, IKEv2NotifyMessageTypes),
         XStrLenField("SPI", "", length_from=lambda pkt: pkt.SPIsize),
         ConditionalField(
-            XStrLenField("notify", "", length_from=lambda pkt: pkt.length - 8),
+            XStrLenField("notify", "", length_from=lambda pkt: pkt.length - 8 - pkt.SPIsize),
```
Comment (mspncp) [inline scapy/contrib/ikev2.py:751]: See the analogous calculation for `ISAKMP_payload_Notify` (line 489):

https://github.com/secdev/scapy/blob/047be325fe7b67fc05287632e7b14bc9e5d6701f/scapy/layers/isakmp.py#L480-L490

**Comment by mspncp** [comment]
Sorry mister [codecov](https://github.com/apps/codecov) but your [statistics](https://app.codecov.io/gh/secdev/scapy/pull/4002/indirect-changes?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=secdev) makes absolutely no sense and your verdict insults me, given all the work I put into testing a single line change ;-)

grafik

**Comment by mspncp** [comment]
CI failures seem to be unrelated:

```
 failed ED1D221E 300.00s AS resolvers
 ```

**Comment by mspncp** [comment]
ping?

**Comment by mspncp** [comment]
Last force-push was just a rebase on master.

**Review by gpotter2 — APPROVED** [review]
Thanks a lot for the PR and sorry for the delay

**Pull request merged**